src/dmd/dmangle.d

```diff
@@ -1202,23 +1202,14 @@ extern (C++) void mangleToBuffer(TemplateInstance ti, OutBuffer* buf)
 }
 
 /******************************************************************************
- * Mangle function signatures ('this' qualifier, and parameter types)
- * to check conflicts in function overloads.
- * It's different from fd.type.deco. For example, fd.type.deco would be null
- * if fd is an auto function.
+ * Mangle function signatures to check conflicts in function overloads.
  *
  * Params:
  *    buf = `OutBuffer` to write the mangled function signature to
 *     fd  = `FuncDeclaration` to mangle
  */
 void mangleToFuncSignature(ref OutBuffer buf, FuncDeclaration fd)
 {
-    auto tf = fd.type.isTypeFunction();
-
     scope Mangler v = new Mangler(&buf);
-
-    MODtoDecoBuffer(&buf, tf.mod);
-    foreach (idx, param; tf.parameterList)
-        param.accept(v);
-    buf.writeByte('Z' - tf.parameterList.varargs);
+    v.mangleExact(fd);
```

Comment:
Can you please elaborate why is this change necessary? It seems like this function does not include the return type in the mangled name. This might be reason why the patch allows overloading on return types.

Reply:
The initial bug report was just one examples where this function would cause false positives (see e.g. the second example in the test). Fixing all of tthem lead me to re-implementing most of the mangling logic.

But I'll check if this can be achieved without these changes.

Reply:
Ditched the current approach in favour of actually comparing the `FuncDeclaration`s